statistics/cmsketch_merge.go

```diff
@@ -0,0 +1,187 @@
+	task := &mergeTask{
+		counter:      make(map[hack.MutableString]float64),
+		datumMap:     make(map[hack.MutableString]types.Datum),
+		hists:        hists,
+		isIndex:      isIndex,
+		killed:       killed,
+		loc:          loc,
+		partitionNum: partitionNum,
+		removeVals:   make([][]TopNMeta, partitionNum),
+		taskChan:     make(chan removeTask, 2),
+		topNs:        topNs,
+		version:      version,
+		wg:           new(errgroup.Group),
+	}
+	return task
+}
+
+func (t *mergeTask) collectRemovedTopN() error {
+	defer func() {
+		close(t.taskChan)
+	}()
+	for i, topN := range t.topNs {
+		if atomic.LoadUint32(t.killed) == 1 {
+			return errors.Trace(ErrQueryInterrupted)
+		}
+		if topN.TotalCount() == 0 {
+			continue
+		}
+		for _, val := range topN.TopN {
+			encodedVal := hack.String(val.Encoded)
+			_, exists := t.counter[encodedVal]
+			t.counter[encodedVal] += float64(val.Count)
+			if exists {
+				// We have already calculated the encodedVal from the histogram, so just continue to next topN value.
+				continue
+			}
+			// We need to check whether the value corresponding to encodedVal is contained in other partition-level stats.
+			// 1. Check the topN first.
+			// 2. If the topN doesn't contain the value corresponding to encodedVal. We should check the histogram.
+			for j := 0; j < t.partitionNum; j++ {
+				if atomic.LoadUint32(t.killed) == 1 {
+					return errors.Trace(ErrQueryInterrupted)
+				}
+				if (j == i && t.version >= 2) || t.topNs[j].findTopN(val.Encoded) != -1 {
+					continue
+				}
+				// Get the encodedVal from the hists[j]
+				datum, exists := t.datumMap[encodedVal]
+				if !exists {
+					// If the datumMap does not have the encodedVal datum,
+					// we should generate the datum based on the encoded value.
+					// This part is copied from the function MergePartitionHist2GlobalHist.
+					var d types.Datum
+					if t.isIndex {
+						d.SetBytes(val.Encoded)
+					} else {
+						var err error
+						if types.IsTypeTime(t.hists[0].Tp.GetType()) {
+							// handle datetime values specially since they are encoded to int and we'll get int values if using DecodeOne.
+							_, d, err = codec.DecodeAsDateTime(val.Encoded, t.hists[0].Tp.GetType(), t.loc)
+						} else if types.IsTypeFloat(t.hists[0].Tp.GetType()) {
+							_, d, err = codec.DecodeAsFloat32(val.Encoded, t.hists[0].Tp.GetType())
+						} else {
+							_, d, err = codec.DecodeOne(val.Encoded)
+						}
+						if err != nil {
+							return err
+						}
+					}
+					t.datumMap[encodedVal] = d
+					datum = d
+				}
+				// Get the row count which the value is equal to the encodedVal from histogram.
+				count, _ := t.hists[j].equalRowCount(nil, datum, t.isIndex)
+				if count != 0 {
+					t.counter[encodedVal] += count
+					// Remove the value corresponding to encodedVal from the histogram.
+					t.taskChan <- removeTask{id: j, topN: TopNMeta{Encoded: val.Encoded, Count: uint64(count)}}
```

Comment:
this taskChan has limited size and may drop removeTask if the reader side is slower than writer side.

Reply:
I have refactored this code to use binary search.